### PR TITLE
Add image support

### PR DESCRIPTION
**Pull request by mitchell-then**
Add support for Markdown files that contain images using additions in jesselang/go-confluence to create and update attachments. 

Companion pull request: https://github.com/jesselang/go-confluence/pull/8

### CODE REVIEW

**Review by hashCode1 — APPROVED** [review]

File: pkg/publish.go [hunk 1]
```diff
@@ -97,3 +124,104 @@ func Publish(file string, rootID string, dryRun bool) (id string, err error) {
...
+	fileDir := filepath.Dir(file)
+	imageSrcs := GetImageSrcsFromHTML(content)
+
+	var imageSrcFiles []string
+
+	for _, imageSrc := range imageSrcs {
+		imageSrcPath := filepath.Join(fileDir, imageSrc)
+		if _, err := os.Stat(imageSrcPath); !os.IsNotExist(err) {
+			imageSrcFiles = append(imageSrcFiles, imageSrc)
+		}
```
Comment (jesselang) [inline pkg/publish.go:185]: What can/should `dox` do when the image file does not exist?

File: pkg/publish.go [hunk 2]
```diff
@@ -97,3 +124,104 @@ func Publish(file string, rootID string, dryRun bool) (id string, err error) {
...
+
+			if fileSum != attachmentSum {
+				_, err := wiki.UpdateAttachment(pageID, imageSrcPath, results.Results[0].ID)
+				if err != nil {
+					return "", err
+				}
+			}
+		}
+
+		pageContent = strings.Replace(pageContent, "\"" + imageSrcFile + "\"", "\"" + uri + "/download/attachments/" + pageID + "/" + imageSrcFilename + "\"", -1)
```
Comment (jesselang) [inline pkg/publish.go]: Consider using back-tick strings to enhance readability when they contain double-quote literals.
```
fmt.Sprintf(`"%s"`, imageSrcFile)
```

File: pkg/publish.go [hunk 3]
```diff
@@ -77,16 +82,38 @@ func Publish(file string, rootID string, dryRun bool) (id string, err error) {
...
+
+			c.Version.Number += 1
+			c, err = wiki.UpdateContent(c)
+			if err != nil {
+				return "", err
+			}
+		}
+
 	} else {
+		newPageContent := src.Output()
```
Comment (jesselang) [inline pkg/publish.go]: This seems to be more aptly named `sourceOutput`, and could be defined closer to where it is used.

File: pkg/publish.go [hunk 4]
```diff
@@ -77,16 +82,38 @@ func Publish(file string, rootID string, dryRun bool) (id string, err error) {
...
 			[]string{"body.storage", "space", "version"})
 		if err != nil {
 			// TODO: handle 404 where dox id exists in source, but published page does not
 			return "", err
 		}
 
-		if c.Body.Storage.Value != src.Output() {
-			c.Body.Storage.Value = src.Output()
+		imageSrcFiles := GetImageSrcFiles(newPageContent, file)
+		newPageContent, err = ReplaceImagesWithAttachments(imageSrcFiles, file, newPageContent, c.ID, wiki, uri)
```
Comment (jesselang) [inline pkg/publish.go]: I'd prefer `newPageContent` being named `pageContent`.

File: pkg/publish.go [hunk 5]
```diff
@@ -97,3 +124,104 @@ func Publish(file string, rootID string, dryRun bool) (id string, err error) {
...
+
+func GetImageSrcsFromHTML(content string) []string {
+	var imageSrcs []string
+	z := html.NewTokenizer(strings.NewReader(content))
+	for {
+		tt := z.Next()
+		t := z.Token()
+		switch tt {
+		case html.ErrorToken:
+			return imageSrcs
```
Comment (jesselang) [inline pkg/publish.go]: If t.Err() is not `io.EOF`, we'll want to return that `error` here (malformed HTML, etc.).

**Review by jesselang — CHANGES_REQUESTED** [review]
Nice work! It looks like none of these added functions need to be exported from the package since they are only used in the package.

File: pkg/publish.go (hunk 1 above)
Comment (mitchell-then) [reply pkg/publish.go:185]: I cheat a lot here by simply checking if an image src exists. If it does, its an image file to be attached. If it doesn't, its likely to be a URL. As you say, there is also a chance that a user wants an image to be included, but it doesn't exist. 

I'll look into making this better. If we can sort out what looks like paths (or URLs), then we can send a warning if the file doesn't exist. 

- [x] make this better

File: pkg/publish.go (hunk 2 above)
Comment (mitchell-then) [reply pkg/publish.go]: - [x] use backticks for strings with `"`

File: pkg/publish.go (hunk 3 above)
Comment (mitchell-then) [reply pkg/publish.go]: - [x] change variable name

File: pkg/publish.go (hunk 4 above)
Comment (mitchell-then) [reply pkg/publish.go]: - [x] change variable name

File: pkg/publish.go (hunk 5 above)
Comment (mitchell-then) [reply pkg/publish.go]: - [x] return `t.Err()` when not equal to `io.EOF`

**Review by jesselang — APPROVED** [review]

**Comment by jesselang** [comment]
Thanks Mitchell!

**Pull request merged**